flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/extraction/TypeInferenceExtractorTest.java

```diff
@@ -0,0 +1,679 @@
+			this.expectedOutputStrategies.put(validator, outputStrategy);
+			return this;
+		}
+
+		TestSpec expectErrorMessage(String expectedErrorMessage) {
+			this.expectedErrorMessage = expectedErrorMessage;
+			return this;
+		}
+	}
+
+	private static class DataTypeLookupMock implements DataTypeLookup {
+
+		@Override
+		public Optional<DataType> lookupDataType(String name) {
+			return Optional.of(TypeConversions.fromLogicalToDataType(LogicalTypeParser.parse(name)));
+		}
+
+		@Override
+		public Optional<DataType> lookupDataType(UnresolvedIdentifier identifier) {
+			return Optional.empty();
+		}
+
+		@Override
+		public DataType resolveRawDataType(Class<?> clazz) {
+			return null;
+		}
+	}
+
+	// --------------------------------------------------------------------------------------------
+	// Test classes for extraction
+	// --------------------------------------------------------------------------------------------
+
+	@FunctionHint(
+		input = {@DataTypeHint("INT"), @DataTypeHint("STRING")},
+		argumentNames = {"i", "s"},
+		output = @DataTypeHint("BOOLEAN")
+	)
+	private static class FullFunctionHint extends ScalarFunction {
+		public Boolean eval(Integer i, String s) {
+			return null;
+		}
+	}
+
+	private static class ComplexFunctionHint extends ScalarFunction {
+		@FunctionHint(
+			input = {@DataTypeHint("INT"), @DataTypeHint(inputGroup = InputGroup.ANY)},
+			argumentNames = {"myInt", "myAny"},
+			output = @DataTypeHint("BOOLEAN"),
+			isVarArgs = true
+		)
+		public Boolean eval(Object... o) {
+			return null;
+		}
+	}
+
+	@FunctionHint(input = @DataTypeHint("INT"), output = @DataTypeHint("INT"))
+	@FunctionHint(input = @DataTypeHint("BIGINT"), output = @DataTypeHint("BIGINT"))
+	private static class FullFunctionHints extends ScalarFunction {
+		public Number eval(Number n) {
+			return null;
+		}
+	}
+
+	@FunctionHint(output = @DataTypeHint("INT"))
+	private static class GlobalOutputFunctionHint extends ScalarFunction {
+		@FunctionHint(input = @DataTypeHint("INT"))
+		public Integer eval(Integer n) {
+			return null;
+		}
+
+		@FunctionHint(input = @DataTypeHint("STRING"))
+		public Integer eval(String n) {
+			return null;
+		}
+	}
+
+	@FunctionHint(output = @DataTypeHint("INT"))
+	private static class InvalidSingleOutputFunctionHint extends ScalarFunction {
+		@FunctionHint(output = @DataTypeHint("STRING"))
+		public Integer eval(Integer n) {
+			return null;
+		}
+	}
+
+	@FunctionHint(input = @DataTypeHint("INT"), output = @DataTypeHint("INT"))
+	private static class SplitFullFunctionHints extends ScalarFunction {
+		@FunctionHint(input = @DataTypeHint("BIGINT"), output = @DataTypeHint("BIGINT"))
+		public Number eval(Number n) {
+			return null;
+		}
+	}
+
+	@FunctionHint(input = @DataTypeHint("INT"), output = @DataTypeHint("INT"))
+	private static class InvalidFullOutputFunctionHint extends ScalarFunction {
+		@FunctionHint(input = @DataTypeHint("INT"), output = @DataTypeHint("BIGINT"))
```

Comment:
Could we make sure the method return types are compatible with the conversion classes in those cases?

It would simplify reasoning. Right now even if we remove one of the output declarations it still fails because the return type mismatches with the declared output type.